src/cylinder_position_filter.cpp

```diff
@@ -0,0 +1,333 @@
+            "that-direction.");
+
+    Nx_ = static_cast<size_t>(1);
+    inv_pitch_x_ = 0.;
+  } else if (pitch_x_ != 0. && Nx_ == 1) {
+    warning(str_pitch_x +
+            "-direction pitch is non-zero, but the bin is one in that "
+            "direction,so, Only one will be considered in "
+            "that-direction.");
+  } else {
+    if (pitch_x_ != 0. && pitch_x_ < (2. * radius_))
+      fatal_error(str_pitch_x + "-pitch is less than the radius");
+    inv_pitch_x_ = 1.0 / pitch_x_;
+  }
+
+  if ((pitch_y_ == 0.) && (Ny_ != 1)) {
+    warning(str_pitch_y +
+            "-direction pitch is zero, Only one will be considered in "
+            "the direction.");
+
+    Ny_ = static_cast<size_t>(1);
+    inv_pitch_y_ = 0.;
+  } else if (pitch_y_ != 0. && Ny_ == 1) {
+    warning(str_pitch_y +
+            "-direction pitch is non-zero, but the bin is one in that "
+            "direction,so, Only one will be considered in "
+            "that-direction.");
+  } else {
+    if (pitch_y_ != 0. && pitch_y_ < (2. * radius_))
+      fatal_error(str_pitch_y + "pitch is less than the radius ");
+    inv_pitch_y_ = 1.0 / pitch_y_;
+  }
+
+  // A finite length the z-direction should be given.
+  if (dz_ == 0.)
+    fatal_error("Length of the cylinder shouldn't zero.");
+  else
+    inv_dz_ = 1.0 / dz_;
+
+  // calculate low point for the purpose of getting the indices
+  // r_low_ is mapped, means z-cooridnate will always be axial direction
+  double low_x = origin_.x() - pitch_x_ * 0.5;
+  double low_y = origin_.y() - pitch_y_ * 0.5;
+  r_low_ = Position(low_x, low_y, origin_.z());
+}
+
+StaticVector3 CylinderPositionFilter::get_indices(const Tracker& tktr) {
+  const Position r = map_coordinate(tktr.r());
+  int nx = 0, ny = 0;
+
+  if (pitch_x_ != 0.)
+    nx = static_cast<int>(std::floor((r.x() - r_low_.x()) * inv_pitch_x_));
+
+  if (pitch_y_ != 0.)
+    ny = static_cast<int>(std::floor((r.y() - r_low_.y()) * inv_pitch_y_));
+
+  int nz = static_cast<int>(std::floor((r.z() - r_low_.z()) * inv_dz_));
+
+  double new_origin_x = origin_.x() + pitch_x_ * static_cast<double>(nx);
+  double new_origin_y = origin_.y() + pitch_y_ * static_cast<double>(ny);
+
+  StaticVector3 indices;
+  // check if nx, ny, and nz are positive
+  // and less the number of bins in that direction
+  if (nx >= 0 && nx < static_cast<int>(Nx_) && ny >= 0 &&
+      ny < static_cast<int>(Ny_) && nz >= 0 && nz < static_cast<int>(Nz_)) {
+    // check if the position is inside the circular radius or not
+    if (sqrt((new_origin_x - r.x()) * (new_origin_x - r.x()) +
+             (new_origin_y - r.y()) * (new_origin_y - r.y())) <=
+        (radius_ + 1E-15)) {
+      indices.push_back(static_cast<std::size_t>(nx));
+      indices.push_back(static_cast<std::size_t>(ny));
+      indices.push_back(static_cast<std::size_t>(nz));
+      map_indexes(indices);
+    } else {
+      return indices;
+    }
+  } else {
+    return indices;
+  }
```

Comment:
```suggestion
  if (nx >= 0 && nx < static_cast<int>(Nx_) && ny >= 0 &&
      ny < static_cast<int>(Ny_) && nz >= 0 && nz < static_cast<int>(Nz_)) {
    // check if the position is inside the circular radius or not
    if (sqrt((new_origin_x - r.x()) * (new_origin_x - r.x()) +
             (new_origin_y - r.y()) * (new_origin_y - r.y())) <=
        (radius_ + 1E-15)) {
      indices.push_back(static_cast<std::size_t>(nx));
      indices.push_back(static_cast<std::size_t>(ny));
      indices.push_back(static_cast<std::size_t>(nz));
      map_indexes(indices);
    }
  }
```
I don't think we need the else statements. In the event we don't enter the if, indices will remain empty.